File: mupattern-py/src/mupattern_py/common/slices.py
from __future__ import annotations


def parse_slice_string(s: str, length: int) -> list[int]:
    """Parse slice expressions like 'all', '1,3', '0:10:2'."""
    if s.strip().lower() == "all":
        return list(range(length))

    indices: set[int] = set()
    for segment in s.split(","):
        segment = segment.strip()
        if not segment:
            continue
        try:
            if ":" in segment:
                parts = [(int(p) if p else None) for p in segment.split(":")]
                if len(parts) == 3 and parts[2] == 0:
                    raise ValueError(f"Slice step cannot be zero: {segment!r}")
                indices.update(range(*slice(*parts).indices(length)))
            else:
                idx = int(segment)
                if idx < -length or idx >= length:
                    raise ValueError(f"Index {idx} out of range for length {length}")
                indices.add(idx % length)
        except ValueError as e:
            if "out of range" in str(e) or "cannot be zero" in str(e):
                raise
            raise ValueError(f"Invalid slice segment: {segment!r}") from e

    if not indices:
        raise ValueError(f"Slice string {s!r} produced no indices")

    return sorted(indices)

File: mupattern-rs/src/slices.rs
/// Parse slice expressions like "all", "1,3", "0:10:2".
/// Semantics mirror muapplication/common/slices.py (slice.indices).
pub fn parse_slice_string(s: &str, length: usize) -> Result<Vec<usize>, String> {
    let s = s.trim();
    if s.eq_ignore_ascii_case("all") {
        return Ok((0..length).collect());
    }

    let len = length as isize;
    let mut indices = std::collections::HashSet::new();

    for segment in s.split(',') {
        let segment = segment.trim();
        if segment.is_empty() {
            continue;
        }
        if segment.contains(':') {
            let parts: Vec<Option<isize>> = segment
                .split(':')
                .map(|p| {
                    let t = p.trim();
                    if t.is_empty() {
                        Ok(None)
                    } else {
                        t.parse()
                            .map_err(|_| format!("Invalid slice segment: {:?}", segment))
                            .map(Some)
                    }
                })
                .collect::<Result<_, _>>()?;
            if parts.len() > 3 {
                return Err(format!("Invalid slice segment: {:?}", segment));
            }
            let start = parts.get(0).copied().flatten().unwrap_or(0);
            let stop = parts.get(1).copied().flatten().unwrap_or(len);
            let step = parts.get(2).copied().flatten().unwrap_or(1);

            if step == 0 {
                return Err(format!("Slice step cannot be zero: {:?}", segment));
            }

            let (start, stop, step) = (start, stop, step);
            let (i, j, k) = slice_indices(start, stop, step, len);
            let mut idx = i;
            while (k > 0 && idx < j) || (k < 0 && idx > j) {
                if idx >= 0 && idx < len {
                    indices.insert(idx as usize);
                }
                idx += k;
            }
        } else {
            let idx: isize = segment
                .parse()
                .map_err(|_| format!("Invalid slice segment: {:?}", segment))?;
            if idx < -len || idx >= len {
                return Err(format!("Index {} out of range", idx));
            }
            let idx = if idx < 0 { idx + len } else { idx };
            indices.insert(idx as usize);
        }
    }

    let mut out: Vec<usize> = indices.into_iter().collect();
    out.sort_unstable();
    Ok(out)
}

/// Mirror Python slice.indices(length) -> (start, stop, step).
fn slice_indices(start: isize, stop: isize, step: isize, length: isize) -> (isize, isize, isize) {
    let (mut start, mut stop) = (start, stop);
    if start < 0 {
        start = (start + length).max(0);
    } else if start > length {
        start = length;
    }
    if stop < 0 {
        stop = (stop + length).max(0);
    } else if stop > length {
        stop = length;
    }
    if step < 0 {
        if stop < start {
            (start, stop.max(-1), step)
        } else {
            (start, -1, step)
        }
    } else {
        if start > stop {
            (start, stop, step)
        } else {
            (start, stop, step)
        }
    }
}
